### ISSUE TITLE
OpenSSL API Compatibility Layer

### ISSUE THREAD

**Issue by SimonTate** [rustls/rustls#1862]
**Checklist**
* [x] I've searched the issue tracker for similar requests

**Is your feature request related to a problem? Please describe.**

Use rustls as a drop in replacement for OpenSSL with the public API from OpenSSL 3.2.0 (or latest stable 3.x version) as per the roadmap (https://github.com/rustls/rustls/blob/main/ROADMAP.md) so that I can use rustls as my main TLS implementation with applications that are not written in rust and depend on openssl.

**Describe the solution you'd like**

libssl.so implemented by rustls with the same header file names and functions as openssl. Allow a `main.c` to be linked with original openssl libs or rustls libs and functionally work the same way. Initially starting by implementing enough to run openssl-bench and then enough to use in curl, and then the whole public API.

**Describe alternatives you've considered**

Convert everyone to using rust 🦀  or could create a separate C library that is NOT a drop in replacement. I think both of these would harm adoption rates though.

**Additional context**

This is part of the roadmap. I'm just creating this for initial discussion on plans to implement before there is a PR opened. Apologies if work has already started on this.

**Comment by djc** [comment]
Work has already started on this -- in private for now. Will discuss if/when we can open it up.

**Comment by SimonTate** [comment]
Good to hear @djc . Propose we leave this open for comment etc when that is opened up. Certainly I'm keen to help on that if I can.

**Comment by mouse07410** [comment]
@djc great to hear! Yes please open this up. Exactly when - decision use yours.

**Comment by djc** [comment]
@SimonTate @mouse07410 can you talk a bit about your respective use cases (assuming they're not the same)? This will help us prioritize our work and consider your perspectives.

**Comment by SimonTate** [comment]
@djc , my use case is quite generic and very broad. I want to be able to replace the libssl.so.3 and associated headers on Linux (I generally use an OpenWrt-based system) with a rustls implementation to allow for use of rustls without changing every application that may use OpenSSL. For me, the aim is to reduce potential CVEs on customer equipment that are often appearing from OpenSSL and C-based TLS implementations.

For example, add support for curl/ wget/ mosquitto/ any higher level TLS application without having to change the code AT ALL from the OpenSSL implementations. It should be zero effort for the maintainers of other packages to use rustls once it is used in the distro they are building for.

It becomes more complicated when other engines are used (like pkcs#11 or whatver) in OpenSSL though, but TLS for curl-like tools without additional application development would be ideal.

**Comment by SimonTate** [comment]
@djc , yeah I'm aware of that and will take more of a look at the issues. Thanks for the reminder 😄 

For this - are you/ people working on this starting with a specific application linking with a rustls based .so? Would be great to build up the functionality of the compatibility layer by enabling applications of increasing complexity (imo).

**Comment by ctz** [comment]
Good question! We have two initial targets: curl and nginx (both unmodified binaries, from ubuntu 22.04 packages). curl works in a basic sense (the dynamic linker is happy and it can fetch google.com); nginx is still in progress.
